src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Framework/ControlFlow/BasicBlock.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.Operations.ControlFlow
+{
+    /// <summary>
+    /// Basic block of a <see cref="ControlFlowGraph"/>.
+    /// NOTE: This class is temporary and will be removed once we move to the CFG exposed from Microsoft.CodeAnalysis
+    /// </summary>
+    [DebuggerDisplay("{Kind} ({Statements.Length} statements)")]
+    internal class BasicBlock
+    {
+        private ImmutableArray<IOperation>.Builder _statements;
+        private ImmutableHashSet<BasicBlock>.Builder _successors;
+        private ImmutableHashSet<BasicBlock>.Builder _predecessors;
+
+        public BasicBlock(BasicBlockKind kind)
+        {
+            Kind = kind;
+            _statements = ImmutableArray.CreateBuilder<IOperation>();
+            _successors = ImmutableHashSet.CreateBuilder<BasicBlock>();
+            _predecessors = ImmutableHashSet.CreateBuilder<BasicBlock>();
+        }
+
+        public BasicBlockKind Kind { get; private set; }
+        public ImmutableArray<IOperation> Statements => _statements.ToImmutable();
+        public ImmutableHashSet<BasicBlock> Successors => _successors.ToImmutable();
```

Comment:
Just curious, why is an `ImmutableHashSet` necessary here?

Reply:
I believe Edgardo's original prototype could invoke AddSuccessor with same block multiple times. This is temporary code that will go away once we move to compiler's CFG, so I am leaving it as is.

---
In reply to: [162773459](https://github.com/dotnet/roslyn-analyzers/pull/1541#discussion_r162773459) [](ancestors = 162773459)